src/components/tooltip/tooltip.tsx

```diff
@@ -0,0 +1,185 @@
+                style-state={{open}}
+            >
+                <GlobalEvent
+                    resize={this.onWindowResize}
+                    mousedown={this.hide}
+                    touchstart={this.hide}
+                />
+                <div className="tooltip">
+                    {children}
+                    <svg className="tail" height="5" width="10">
+                        <polygon points="0,0 10,0 5,5"/>
+                    </svg>
+                </div>
+            </div>
+        );
+    }
+
+    public componentDidMount() {
+        this.setTarget();
+        this.bindEvents();
+        this.setStyles();
+    }
+    public componentWillUnmount() {
+        window.clearTimeout(this.timeout!);
+        this.unbindEvents();
+        this.onWindowResize.clear();
+    }
+    public componentWillReceiveProps(props: TooltipProps) {
+        this.unbindEvents();
+        this.setTarget();
+        this.bindEvents();
+        this.setStyles();
+    }
+
+    private setTarget() {
+        this.target = document.querySelector(`[data-tooltip-for=${this.props.id}]`) as HTMLElement;
+    }
+
+    private bindEvents() {
+        if (!this.target) {
+            return;
+        }
+        const {showTrigger, hideTrigger} = this.props;
+        this.events = showTrigger!.split(',')
+            .concat(hideTrigger!.split(','))
+            .filter((val, index, arr) => arr.indexOf(val) === index);
+        this.events.forEach(event => {
+            this.target!.addEventListener(event, this.toggle);
+        });
+    }
+    private unbindEvents() {
+        if (!this.target) {
+            return;
+        }
+        this.events.forEach(event => {
+            this.target!.removeEventListener(event, this.toggle);
+        });
+    }
+    private setStyles() {
+        if (!this.target) {
+            return;
+        }
+        const {position} = this.props;
+        const rect = this.target!.getBoundingClientRect();
+        const top = rect.top + window.scrollY;
+        const left = rect.left + window.scrollX;
+        const style = {
+            top: position === 'bottom' ? (top + rect.height) : top,
+            left: position === 'right' ? (left + rect.width) : left,
+            marginLeft: (position === 'top' || position === 'bottom') ? (rect.width / 2) : undefined,
+            marginTop: (position === 'left' || position === 'right') ? (rect.height / 2) : undefined
+        };
+        this.setState({style});
+    }
+
+    private toggle = (e: Event) => {
+        const {open} = this.state;
+        const {showTrigger, hideTrigger, showDelay, hideDelay} = this.props;
+        const {type} = e;
+        const delay = open ? hideDelay : showDelay;
+
+        const next = () => {
+            if (
+                (open && hideTrigger!.indexOf(type) !== -1) ||
+                (!open && showTrigger!.indexOf(type) !== -1)
+            ) {
+                this.setState({open: !open});
+            }
+        };
```

Comment:
maybe we can avoid creating function in event handler? put in separate method for eaxmple

Reply:
we can't do anything better here, because we have `type` in closure and we will need to pass it to another function wherever it will be 
```ts
private immediateToggle(type) { ...}
private toggle(e) {
  const {type} = e;
  if (delay) {
     // extra function here
     this.timer = window.setTimeout(() => this.immediateToggle(type), delay)
  } else {
     this.immediateToggle(type);
  }
}
```